File: backend/app/models/progression.py
"""Progression data models."""


from pydantic import BaseModel


class ProgressionSet(BaseModel):
    """One logged set within a day."""
    weight: float
    weight_unit: str
    reps: int
    estimated_1rm: float
    comment: str | None = None


class ProgressionDataPoint(BaseModel):
    """A day's work on one exercise.

    The top-level numbers are the day's *best* set by estimated 1RM, which is
    what the chart plots — one point per session, because plotting every set
    turns a progression line into a cloud. `sets` carries the whole day
    underneath it, so the history list can show the session as it happened:
    three sets across at 160, not just the heaviest of them.
    """
    date: str
    weight: float
    weight_unit: str
    reps: int
    estimated_1rm: float
    comment: str | None = None
    sets: list[ProgressionSet] = []


class UpcomingProgressionDataPoint(BaseModel):
    """Single upcoming progression data point."""
    session: int
    projected_date: str
    weight: float
    weight_unit: str
    reps: int
    estimated_1rm: float
    comment: str | None = None


class ProgressionResponse(BaseModel):
    """Progression data for an exercise."""
    exercise: str
    historical: list[ProgressionDataPoint]
    upcoming: list[UpcomingProgressionDataPoint]

File: frontend/src/types/progression.ts
export interface ProgressionSet {
  weight: number;
  weight_unit: string;
  reps: number;
  estimated_1rm: number;
  comment: string | null;
}

/**
 * A day's work on one exercise. The top-level numbers are the day's best set
 * by estimated 1RM — the single point the chart plots — and `sets` carries the
 * whole session underneath, in the order performed.
 */
export interface ProgressionDataPoint {
  date: string;
  weight: number;
  weight_unit: string;
  reps: number;
  estimated_1rm: number;
  comment: string | null;
  sets: ProgressionSet[];
}

export interface UpcomingProgressionDataPoint {
  session: number;
  projected_date: string;
  weight: number;
  weight_unit: string;
  reps: number;
  estimated_1rm: number;
  comment: string | null;
}

export interface ProgressionResponse {
  exercise: string;
  historical: ProgressionDataPoint[];
  upcoming: UpcomingProgressionDataPoint[];
}
